pi_top_usb_setup/pages/run_setup.py

```diff
@@ -0,0 +1,200 @@
+import logging
+import os
+from enum import Enum
+from pathlib import Path
+from threading import Thread
+
+from pitop.common.command_runner import run_command
+from pt_miniscreen.components.mixins import Actionable
+from pt_miniscreen.components.progress_bar import ProgressBar
+from pt_miniscreen.core.component import Component
+from pt_miniscreen.core.components import Text
+from pt_miniscreen.core.utils import apply_layers, layer
+
+from pi_top_usb_setup.common import (
+    MOUNT_POINT,
+    OFFLINE_APT_SOURCE,
+    UPDATES_FOLDER,
+    USB_SETUP_FILE,
+)
+from pi_top_usb_setup.utils import (
+    close_app,
+    drive_has_enough_free_space,
+    extract_file,
+    get_extracted_size,
+    get_package_version,
+)
+
+logger = logging.getLogger(__name__)
+
+
+FONT_SIZE = 14
+
+
+class RunStates(Enum):
+    ERROR = -1
+    INIT = 0
+    EXTRACTING_TAR = 1
+    UPDATING_SOURCES = 2
+    INSTALLING_UPDATES = 3
+    DONE = 4
+
+
+class AppErrors(Enum):
+    NONE = 0
+    NOT_ENOUGH_SPACE = 1
+    UPDATE_ERROR = 2
+
+
+class RunSetupPage(Component, Actionable):
+    def __init__(self, **kwargs):
+        super().__init__(
+            initial_state={"run_state": RunStates.INIT, "error": AppErrors.NONE},
+            **kwargs,
+        )
+
+        self.text_component = self.create_child(
+            Text,
+            text=self._text(),
+            get_text=self._text,
+            font_size=FONT_SIZE,
+            align="center",
+            vertical_align="center",
+            wrap=False,
+        )
+        self.progress_bar = self.create_child(
+            ProgressBar, progress=self._current_progress
+        )
+
+        Thread(target=self.run_setup, daemon=True).start()
+
+    def _text(self):
+        run_state = self.state.get("run_state")
+        if run_state == RunStates.DONE:
+            return "Device setup\nis complete;\nPress the select\nbutton to exit ..."
+        elif run_state == RunStates.ERROR:
+            error = self.state.get("error")
+            if error == AppErrors.NOT_ENOUGH_SPACE:
+                return "USB drive doesn't\nhave enough free\nspace"
+            return "There was an error,\n please try again"
+        return "Please wait ..."
+
+    def run_setup(self):
+        if not Path(USB_SETUP_FILE).exists():
+            logger.warning(f"File {USB_SETUP_FILE} not found, exiting ...")
+            close_app()
+
+        # Check if there's enough free space
+        if not drive_has_enough_free_space(
+            drive=MOUNT_POINT, space=get_extracted_size(USB_SETUP_FILE)
+        ):
+            self.state.update(
+                {"run_state": RunStates.ERROR, "error": AppErrors.NOT_ENOUGH_SPACE}
+            )
+            return
+
+        # Extract compressed file
+        self.state.update({"run_state": RunStates.EXTRACTING_TAR})
+        extract_file(file=USB_SETUP_FILE, destination=MOUNT_POINT)
```

Comment:
```suggestion
        extract_file(file=USB_SETUP_FILE, destination=MOUNT_POINT)
```
probably this should be try-catched